src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RandomNumberGenerator.cs

```diff
@@ -350,6 +353,44 @@ private static void GetHexStringCore(Span<char> destination, bool lowercase)
 
         private static void GetItemsCore<T>(ReadOnlySpan<T> choices, Span<T> destination)
         {
+            // The most expensive part of this operation is the call to get random data. We can
+            // do so only once rather than per element in a common case:
+            // - The number of choices is <= 256. This let's us get a single byte per choice.
+            // - The number of choices is a power of two. This let's us use a byte and simply mask off
+            //   unnecessary bits cheaply rather than needing to use rejection sampling.
+            if (BitOperations.IsPow2(choices.Length) && choices.Length <= 256)
+            {
+                const int StackAllocThreshold = 256;
+
+                // Get a scratch buffer to fill with random bytes, either stack space if the
+                // destination is small enough, or an array pool array if it's larger.
+                // If T is unmanaged, we _could_ use the destination buffer itself, reinterpreted
+                // as a byte buffer, to hold the scratch bytes, but then we'd potentially be temporarily
+                // writing invalid Ts into the destination, which is a risk.
+                byte[]? arrayPoolArray = null;
+                Span<byte> randomBytes = destination.Length <= StackAllocThreshold ?
+                    stackalloc byte[StackAllocThreshold] :
+                    (arrayPoolArray = ArrayPool<byte>.Shared.Rent(destination.Length));
+                randomBytes = randomBytes.Slice(0, destination.Length);
+
+                Fill(randomBytes);
```

Comment:
It will probably get inlined, but we can go directly to the internal implementation here.

```suggestion
                RandomNumberGeneratorImplementation.FillSpan(randomBytes);
```